contracts/rbn-staking/IncentivizedVotingLockup.sol

```diff
@@ -0,0 +1,812 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import {IIncentivisedVotingLockup} from "../interfaces/IIncentivisedVotingLockup.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {IERC20Detailed} from "../interfaces/IERC20Detailed.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import {StableMath} from "../libraries/StableMath.sol";
+import {Root} from "../libraries/Root.sol";
+
+/**
+ * @title  IncentivisedVotingLockup
+ * @author Voting Weight tracking & Decay
+ *             -> Curve Finance (MIT) - forked & ported to Solidity
+ *             -> https://github.com/curvefi/curve-dao-contracts/blob/master/contracts/VotingEscrow.vy
+ *         osolmaz - Research & Reward distributions
+ *         alsco77 - Solidity implementation
+ * @notice Lockup MTA, receive vMTA (voting weight that decays over time), and earn
+ *         rewards based on staticWeight
+ * @dev    Supports:
+ *            1) Tracking MTA Locked up (LockedBalance)
+ *            2) Pull Based Reward allocations based on Lockup (Static Balance)
+ *            3) Decaying voting weight lookup through CheckpointedERC20 (balanceOf)
+ *            5) Migration of points to v2 (used as multiplier in future) ***** (rewardsPaid)
+ *            6) Closure of contract (expire)
+ */
+contract IncentivisedVotingLockup is
+  IIncentivisedVotingLockup,
+  ReentrancyGuard
+{
+  using StableMath for uint256;
+  using SafeERC20 for IERC20;
+
+  /** Shared Events */
+  event Deposit(
+    address indexed provider,
+    uint256 value,
+    uint256 locktime,
+    LockAction indexed action,
+    uint256 ts
+  );
+  event Withdraw(address indexed provider, uint256 value, uint256 ts);
+
+  // RBN Redeemer contract
+  address public rbnRedeemer;
+
+  /** Shared Globals */
+  IERC20 public stakingToken;
+  uint256 private constant WEEK = 7 days;
+  uint256 public constant MAXTIME = 4 * 365 days; // 4 years
+  uint256 public END;
+
+  /** Lockup */
+  uint256 public globalEpoch;
+  uint256 public totalLocked;
+  Point[] public pointHistory;
+  mapping(address => Point[]) public userPointHistory;
+  mapping(address => uint256) public userPointEpoch;
+  mapping(uint256 => int128) public slopeChanges;
+  mapping(address => LockedBalance) public locked;
+
+  // Voting token - Checkpointed view only ERC20
+  string public name;
+  string public symbol;
+  uint256 public decimals;
+
+  /** Structs */
+  struct Point {
+    int128 bias;
+    int128 slope;
+    uint256 ts;
+    uint256 blk;
```

Comment:
You can compact these to `uint128` so that it fits into a single slot. Should work the same because timestamp and block will never exceed 128